[torchelastic] Don't do signal handling when off the main thread

evenlion opened the pull request:
Summary:
In multiprocessing, signal handling is not possible if the thread is not the main thread. This resulted in the following error:
> "ValueError('signal only works in main thread of the main interpreter')"

To address this issue, the diff checks whether the thread is the main thread and, if not, skips signal handling.

Test Plan:
Before this change, MAST job failed:
https://fburl.com/mlhub/iq2m10v8

With this change, MAST job succeeded:
https://fburl.com/mlhub/q6kb8343

Differential Revision: D62166943


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

d4l3k commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -470,11 +471,12 @@ def __init__(
+            signal.signal(signal.SIGINT, _terminate_process_handler)
+            if not IS_WINDOWS:
+                signal.signal(signal.SIGHUP, _terminate_process_handler)
+                signal.signal(signal.SIGQUIT, _terminate_process_handler)
```
This has a lot of potential to cause issues since child processes won't be correctly cleaned up without this.

Is this being called programmatically? Can we pull this logic out into a helper function that can be explicitly called from the main thread so we keep this logic as well as support non-main threads?

d4l3k replied:
Actually reading this more -- we need torchelastic to run in the main thread as these handlers only run in the main thread

d4l3k commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -470,11 +471,12 @@ def __init__(
-        if not IS_WINDOWS:
-            signal.signal(signal.SIGHUP, _terminate_process_handler)
-            signal.signal(signal.SIGQUIT, _terminate_process_handler)
+        if threading.current_thread() is threading.main_thread():
```
We should have a proper error message/warning when something like this occurs since it can cause unexpected behavior.

d4l3k commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -470,11 +471,12 @@ def __init__(
+            signal.signal(signal.SIGINT, _terminate_process_handler)
+            if not IS_WINDOWS:
+                signal.signal(signal.SIGHUP, _terminate_process_handler)
+                signal.signal(signal.SIGQUIT, _terminate_process_handler)
```
We should also add some tests for this behavior

d4l3k replied:
Would be nice to add a test to verify elastic can work on a child thread to avoid regressions

There's some similar tests in https://github.com/pytorch/pytorch/blob/ed06772e350c5d550e6a2212d01f5379db475bf3/test/distributed/elastic/multiprocessing/api_test.py#L350-L363 which you can base it off of

d4l3k commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -470,11 +471,16 @@ def __init__(
+                signal.signal(signal.SIGQUIT, _terminate_process_handler)
+        else:
+            logger.warning(
+                "Off the main thread. This could lead to orphaned child processes if the torchrun is terminated."
```
nit: "Failed to register signal handlers since torchelastic is running on a child thread. This could lead to orphaned worker processes if the torchrun is terminated."

d4l3k reviewed (approved):
LGTM

evvens committed:
[torchelastic] Don't do signal handling when off the main thread (#135088)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/135088

In multiprocessing, signal handling is not possible if the thread is not the main thread. This resulted in the following error:
> "ValueError('signal only works in main thread of the main interpreter')"

To address this issue, the diff checks whether the thread is the main thread and, if not, skips signal handling.

Test Plan:
Added a new unit test ```test_pcontext_wait_on_a_child_thread```
  buck2 test mode/dev-nosan //caffe2/test/distributed/elastic/multiprocessing:api_test -- --exact 'caffe2/test/distributed/elastic/multiprocessing:api_test - test_pcontext_wait_on_a_child_thread (api_test.ForkServerTest)'

Without the change:
Test failed with
> ValueError: signal only works in main thread of the main interpreter

With the change:
Test passed

Reviewed By: d4l3k

Differential Revision: D62166943
